### ISSUE TITLE
os/signal.Notfiy: filter out runtime-generated SIGURG?

### ISSUE THREAD

**Issue by minux** [golang/go#37942]
This program, when running under go 1.14, will show a huge number of `SIGURG` are received even though none should be (no network IO, no explicit kills).

https://play.golang.org/p/x7hFjPZnrg5
<pre>
package main

import (
	"fmt"
	"os"
	"os/signal"
	"syscall"
	"time"
)

func main() {
	ch := make(chan os.Signal, 1)
	signal.Notify(ch, syscall.SIGURG)

	for {
		select {
		case sig := <-ch:
			fmt.Printf("received %v: %s\n", sig, time.Now())
		default:
			_ = new(int) // generate some GC activities
		}
	}
}
</pre>

https://go.googlesource.com/proposal/+/master/design/24543-non-cooperative-preemption.md changes to use SIGURG to preempt goroutines, which is fine, but I think `os/signal.Notify` should filter out those runtime generated SIGURGs.

The reason I found this is that I have a program that happens to use `SIGURG` as a custom signal like `SIGUSR1`, which works fine with previous Go releases. I understand that it should probably use `SIGUSR1`, but still I think `os/signal` should hide any signals generated by the runtime as it's irrelevant for the user and an implementation detail.

/cc @aclements thoughts?

**Comment by changkun** [comment]
I got the roughly same issue regarding `SIGURG`. Say this example:

```diff
package main

import (
	"fmt"
	"os"
	"os/signal"
	"runtime"
	"syscall"
)

func main() {
	s := make(chan os.Signal, 1)
	signal.Notify(s)
	go func() { // mocking preemption
		for {
		}
	}()
	for {
		switch signal := <-s; signal {
		case os.Interrupt, os.Kill, syscall.SIGABRT, syscall.SIGTERM:
			fmt.Println("stopped")
			return
+		case syscall.SIGURG:
+			runtime.Gosched()
		default:
                         // .. some work for unhandled signals, do not expect SIGURG ..
			fmt.Printf("unhandled signal: %v\n", signal)
		}
	}
}
```

where a program before 1.14 doesn't expect `SIGURG`, but has to take care of `SIGURG` and call `runtime.Gosched()` go avoid tight loop.

This is kind of incompatible of the previous Go code (code behavior change).

**Comment by aclements** [comment]
> The reason I found this is that I have a program that happens to use SIGURG as a custom signal like SIGUSR1

Right. It was an explicit decision to pass these through to the program, in case the program is using SIGURG. Unfortunately, the runtime can't tell if a signal is "runtime generated" and thus can't filter it out. Because the kernel combines signals, it's not even well-specified what it would mean for a received signal to be runtime-generated because the kernel could have combined a runtime-generated signal with a signal from another source.

@changkun, I don't really understand your program. First, I don't understand the need for `Gosched` there. There's no tight loop, and even if there were, I don't see what that `Gosched` would accomplish. Second, if a program is going to subscribe to all signals, rather than passing specific signals of interest to `signal.Notify`, then it should be prepared to receive any signal. There are lots of reasons lots of signals get sent. For example, if I ran that program in a terminal emulator and started resizing the window, it would begin streaming warnings about unexpected SIGWINCHes.

**Comment by changkun** [comment]
@aclements I hope I could make my point clear this time.

First, this loop is in my program somewhere, works as expected. Completely silence:

```
	for {
		switch signal := <-s; signal {
		case os.Interrupt, os.Kill, syscall.SIGABRT, syscall.SIGTERM:
			fmt.Println("stopped")
			return
		default:
                         // .. some work for unhandled signals, do not expect SIGURG ..
			fmt.Printf("unhandled signal: %v\n", signal)
		}
	}
```

After upgrade to 1.14, the `default` branch start gets many `SIGURG` signals, which is unexpected before. Therefore, a fix must be patched:

```
	for {
		switch signal := <-s; signal {
		case os.Interrupt, os.Kill, syscall.SIGABRT, syscall.SIGTERM:
			fmt.Println("stopped")
			return
+		case syscall.SIGURG:
+			runtime.Gosched()
		default:
                         // .. some work for unhandled signals, do not expect SIGURG ..
			fmt.Printf("unhandled signal: %v\n", signal)
		}
	}
```

for the ignoring of unexpected `SIGURG` signal in my program. You can argue that the user should responsible for all possible incoming signals, but if it doesn't appear before, then I consider this is a behavior change.

Do not get confused by the `runtime.Gosched()`, it means nothing (well, maybe a user may guess `SIGURG` massively appears, then its case will become a tight loop branch. Note that I am just saying and knowing that the preemption signalling is not tight).

**Comment by ianlancetaylor** [comment]
@minux I don't see how it is possible even in principle for os/signal to filter out signals generated from the runtime package.  Do you have any suggestions as to how that could be done?

**Comment by cherrymui** [comment]
`signal.Notify` is to capture process-targeted signal, right? The preemption signals are thread-targeted. Maybe we can filter signals that `sigcode` is `SI_TKILL` (at least on Linux)?

**Comment by cherrymui** [comment]
That would not work if the kernel combines signals with different `sigcode`. Does it?

**Comment by ianlancetaylor** [comment]
As far as I can tell, the Linux 5.3.12 kernel does combine signals with a different `si_code` value.  In `__send_signal` I see

```C
	if (legacy_queue(pending, sig))
		goto ret;
```

and

```C
static inline bool legacy_queue(struct sigpending *signals, int sig)
{
	return (sig < SIGRTMIN) && sigismember(&signals->signal, sig);
}
```

where `signals->signal` is the set of pending signals.

**Comment by minux** [comment]
Yeah, it's very difficult to filter out runtime-generated SIGURG without
also occasionally remove a real SIGURG.

I guess, a slightly better solutions is to make sure the runtime only uses
SIGURG when it's necessary.
My program can tolerate occasional spurious SIGURG, but in the current
runtime implementation, it's occurring too frequently.
(Just like the example program I posted, it contains enough preempt points
that forced preemption shouldn't be necessary.)

Perhaps we can repurpose this issue to "runtime: only use forced preemption
when necessary"? Thoughts?

**Comment by mengzhuo** [comment]
I found a lot of SIGUSR will failed the runtime/regex testes of mips64le due to timeout.

https://farmer.golang.org/temporarylogs?name=linux-mips64le-mengzhuo&rev=dcf0929de6a12103a8fd7097abd6e797188c366d&st=0xc010fd8ea0

**Comment by ianlancetaylor** [comment]
@mengzhuo Can you open a separate issue for that?  It sounds like the program is having a hard time making progress, which seems more like #37741.  Thanks.

**Comment by aclements** [comment]
> Perhaps we can repurpose this issue to "runtime: only use forced preemption when necessary"?

I've considered delaying the first attempt to use signals until some time after we've tried cooperative preemption and it's failed. Most of the time cooperative preemption responds very quickly. I'm not sure what to set that threshold to, though, since every millisecond of delay there is a millisecond added to STW if there's anything not cooperating.

**Comment by networkimprov** [comment]
Could you also provide a build option to enable (or never enable) preemptive? Runtime env flags would presumably override that...

**Comment by aclements** [comment]
That option already exists (`GODEBUG=asyncpreemptoff=1`), but, as it says on the tin, it's meant for debugging.

**Comment by networkimprov** [comment]
I wrote, "... provide a _build_ option ..." e.g. a compiler flag.
